uReplicator-Manager/src/main/java/com/uber/stream/kafka/mirrormaker/manager/core/ControllerHelixManager.java

```diff
@@ -1089,47 +1094,46 @@ public InstanceTopicPartitionHolder createNewRoute(String pipeline, int routeId)
     }
 
     _availableControllerList.remove(instanceName);
-    _pipelineToInstanceMap.put(pipeline, new PriorityQueue<>(1,
+    _pipelineToInstanceMap.putIfAbsent(pipeline, new PriorityQueue<>(1,
         InstanceTopicPartitionHolder.totalWorkloadComparator(_pipelineWorkloadMap)));
     _pipelineToInstanceMap.get(pipeline).add(instance);
     _assignedControllerCount.inc();
     _workerHelixManager.addTopicToMirrorMaker(instance, pipeline, routeId);
 
     // register metrics
-    String routeString = srcDst[1] + "-" + srcDst[2] + "-" + routeId;
     maybeRegisterMetrics(routeString);
 
     spectator.disconnect();
     return instance;
   }
 
   public InstanceTopicPartitionHolder maybeCreateNewRoute(
-      PriorityQueue<InstanceTopicPartitionHolder> instanceList,
-      String topicName,
-      int numPartitions,
-      String pipeline) throws Exception {
+          PriorityQueue<InstanceTopicPartitionHolder> instanceList,
+          String topicName,
+          int numPartitions,
+          String pipeline) throws Exception {
 
     LOGGER.info("maybeCreateNewRoute, topicName: {}, numPartitions: {}, pipeline: {}", topicName, numPartitions,
-        pipeline);
+            pipeline);
 
-    Set<Integer> routeIdSet = new HashSet<>();
+    int routeId = 0;
+    String routeName = ControllerUtils.getRouteName(pipeline, routeId);
     for (InstanceTopicPartitionHolder instance : instanceList) {
-      if (instance.getTotalNumPartitions() + numPartitions < _initMaxNumPartitionsPerRoute) {
-        return instance;
+      if(routeId < instance.getRoute().getPartition()){
+        routeName = instance.getRouteString();
       }
-      routeIdSet.add(instance.getRoute().getPartition());
     }
 
-    // For now we don't delete route even it's empty. so routeId should be 0,1...N
-    int routeId = 0;
-    while (routeId < routeIdSet.size() + 1) {
-      if (!routeIdSet.contains(routeId)) {
-        break;
-      }
-      routeId++;
+    LOGGER.info("maybeCreateNewRoute, routeName: {}", routeName);
+
+    if(_routeToPartitionMap.get(routeName) + numPartitions < _initMaxNumPartitionsPerRoute){
+      _workerHelixManager.expandPipelineInMirrorMaker(pipeline, routeId);
+      ((ControllerLiveInstanceChangeListener)_liveInstanceChangeListener).triggerRebalance();
```

Comment:
there is no need to manually trigger rebalance in there

Reply:
yes,  it's no need now.
but i think it can work well when worker is in high workload, no need to wait a long scheduled rebalance maybe。